autoextract/aio/retry.py

```diff
@@ -64,92 +63,105 @@ def _is_retriable_query_error(exc: Exception) -> bool:
     return isinstance(exc, _QueryError) and exc.retriable and exc.max_retries > 0
 
 
-autoextract_retry_condition = (
-    retry_if_exception(_is_throttling_error) |
-    retry_if_exception(_is_network_error) |
-    retry_if_exception(_is_server_error) |
-    retry_if_exception(_is_retriable_query_error)
-)
-
-
-class autoextract_wait_strategy(wait_base):
-    def __init__(self):
-        # throttling
-        self.throttling_wait = wait_chain(
-            # always wait 20-40s first
-            wait_fixed(20) + wait_random(0, 20),
-
-            # wait 20-40s again
-            wait_fixed(20) + wait_random(0, 20),
+def _exception_factory(fut):
+    exc = fut.exception()
+    if isinstance(exc, _QueryError):
+        return QueryRetryError(fut)
 
-            # wait from 30 to 630s, with full jitter and exponentially
-            # increasing max wait time
-            wait_fixed(30) + wait_random_exponential(multiplier=1, max=600)
-        )
+    return RetryError(fut)
 
-        # connection errors, other client and server failures
-        self.network_wait = (
-            # wait from 3s to ~1m
-            wait_random(3, 7) + wait_random_exponential(multiplier=1, max=55)
-        )
-        self.server_wait = self.network_wait
-        self.retriable_wait = self.network_wait
 
-    def __call__(self, retry_state: RetryCallState) -> float:
+class RetryFactory:
+    """
+    Build custom retry configuration
+    """
+    retry_condition = (
+            retry_if_exception(_is_throttling_error) |
+            retry_if_exception(_is_network_error) |
+            retry_if_exception(_is_server_error) |
+            retry_if_exception(_is_retriable_query_error)
+    )
+    # throttling
+    throttling_wait = wait_chain(
+        # always wait 20-40s first
+        wait_fixed(20) + wait_random(0, 20),
+
+        # wait 20-40s again
+        wait_fixed(20) + wait_random(0, 20),
+
+        # wait from 30 to 630s, with full jitter and exponentially
+        # increasing max wait time
+        wait_fixed(30) + wait_random_exponential(multiplier=1, max=600)
+    )
+
+    # connection errors, other client and server failures
+    network_error_wait = (
+        # wait from 3s to ~1m
+        wait_random(3, 7) + wait_random_exponential(multiplier=1, max=55)
+    )
+    server_error_wait = network_error_wait
+    retriable_query_error_wait = network_error_wait
+    throttling_stop = stop_never
+    network_error_stop = stop_after_delay(15 * 60)
+    server_error_stop = stop_after_delay(15 * 60)
+    retryable_query_error_stop = stop_after_delay(15 * 60)
+
+    def __init__(self, **kwargs):
+        mutable_attributes = {attrib for attrib in dir(self)
+                              if not attrib.startswith("_")}
+        for k, v in kwargs:
+            if not k in mutable_attributes:
+                raise TypeError(f"Attribute {k} is invalid. Any of "
+                                f"{mutable_attributes} was expected")
+            setattr(self, k, v)
```

Comment:
Is this a shortcut to allow setting some of the attributes from the constructor?
If so, it'd be good to have tests for it.
However, maybe we can remove this feature? One can subclass RetryFactory and change an attribute this way, should be easy enough, and won't require code in `__init__`.

Reply:
Sure, sure! I thought I have already removed it. Agree with you, no `__init__` at all is required here. Removed here: https://github.com/scrapinghub/scrapinghub-autoextract/pull/24/commits/62dea1775a3afe8ca136f7e437cebe649b74e689